### PR TITLE
✨ Use PL callback for MLflow use case

### PR DESCRIPTION
**Pull request by Zethson**
We're now using a `LaminDBCheckpointCallback` which creates a new Artifact whenever an epoch ends. This makes it safe to cancel training runs midway as all checkpoints are immediately saved to LaminDB after the respective epoch has completed.

### CODE REVIEW

**Comment by Zethson** [comment]
> I didn't realize this earlier, but it'd be really great to have more than a single run here in the UI, and more than 2 points. 😆
> image
> 
> Because of course it's easy to keep track of a single run, but lamindb shines when you have many runs. So I'd say at least 2 or 3?
> 
> It could also be nice to clarify that "nebolous-..." is the MLFlow run id? (Or name?)

This is the detailed view **of a single run**. I'll replace this screenshot with one that had a few more epochs.

The experiment overview that you were probably thinking of looks like this:

image

**Comment by Zethson** [comment]
@falexwolf 

I had another stab at it and I think that we're converging on something good. I can optimize the screenshots further to make it clearer how MLflow runs and LaminDB runs are connected but before I spend more time on this I think it might be good to have another round of feedback of yours.

I understand the importance of this guide currently but I think that we should stop at a reasonable point before optimizing it further as we might overhaul it completely in a few weeks.

**Comment by falexwolf** [comment]
image

... to track model checkpoints as artifacts with data lineage.

... to track model checkpoints as artifacts linked against training runs.

---

This is difficult to phrase correctly and concisely because MLFlow is a "tracking tool", too. But it doesn't have data lineage and a much narrower scope (at least that's what I understand).

**Comment by falexwolf** [comment]
```python
hyperparameter = ln.Feature(name="Autoencoder hyperparameter", is_type=True).save()
hyperparams = ln.Feature.from_dict(MODEL_CONFIG, str_as_cat=True)
for param in hyperparams:
    param.type = hyperparameter
    param.save()
```

Apologies that I didn't notice this earlier. We have the pattern that would make this nice in the lamindb API already e.g. in `.from_values()`: https://docs.lamin.ai/lamindb.feature#lamindb.Feature.from_values

We add an `organism` and it gets set on every feature.

In an earlier version we had this based on `**kwargs` and then passing `type` would already have worked.

So, I think with 15 min of work, you could extend `from_dict()` to understand a `type` param and this would be really helpful.

---

I think Julianus hit the nail on the head with something that's been quite concerning for a long time ([here](https://github.com/laminlabs/pfizer-lamin-usage/issues/425#issuecomment-3350399855)):

> If I do not need the input params searchable etc, I just want them to be attached as a little json like for nextflow pipelines (or are they features there as well and I did not notice?)

Here, I think there is a rather strong case for querying artifacts and runs by params and having validation for these params across runs. And a one line call to define that validation seems bearable.

The `str_as_cat` though looks scary and I'd remove it.

**Comment by falexwolf** [comment]
I would break this out in two boxes:

image

First: First define features & labels so that strings pass validation later on.

```python
MODEL_CONFIG = {"batch_size": 32, "lr": 0.001}
hyperparam_type = ln.Feature(name="Autoencoder hyperparameter", is_type=True).save()
hyperparams = ln.Feature.from_dict(MODEL_CONFIG, type=hyperparam_type)   # this cannot not yet save; cannot mutate database state in the constructor (we had this discussion a while ago for Schema()) :)
ln.save(hyperparams)
ln.Project(name="MLflow tutorial")
```

Second: Track this notebook or script run so that all checkpoint artifacts come with a link to the source code that created them.

```python
ln.track(params=MODEL_CONFIG, project="MLflow tutorial")
```

**Comment by falexwolf** [comment]
I think this would benefit from a "description".

image

Because it's not just a file with a suffix, but it's a folder that deserves some explanation.

**Comment by falexwolf** [comment]
Given you educated me that `__getattr__` resolves I think this becomes more elegant if you do:

image

```
lamindb_callback = ln.integrations.lightning.LightningCallback()
```

Or potentially even:

```
lamindb_callback = ln.integrations.lightning.Callback()
```

Because there is no confusion with other `Callback` classes, it is scoped to `lamindb.lightning` just like `lightning` itself scopes its `lightning.Callback` class to its root module: `lightning`. ☺️

**Comment by falexwolf** [comment]
I like this!

image

Just what I don't like about it is that the feature definition occurs in the actual training code. If we want to showcase this to the user I'd add those feature definitions at the very top of the notebook:

> First: First define features & labels so that strings pass validation later on.

**Comment by falexwolf** [comment]
Can we get the mlflow run id from `self: Callback`?

If not, we need to actually bother the user with passing it to `Callback()`.

Because we do **not** map mlflow run on lamindb run 1:1 and hence cannot leverage `ln_run.reference` for this, we need to store the mlflow run as a linked feature here: https://github.com/laminlabs/lamindb/pull/3115/files#r2397629852

For managing `MLFlow` runs, I recommend using `Record`, because we'll now have the ability to take notes etc. and it should look very intuitive on the GUI very soon (more intuitive than ulabel).

All of this only has to be defined once and in the future will likely often be done on the UI:
```python
run_type = ln.Record(name="MLFlow runs", is_type=True).save()
my_runs_this_week = ln.Record(name="My runs week 3", is_type=True, type=mlflow_type).save()
ln.Feature(name="mlflow_run_id", dtype=run_type).save()
```

I'd add a screenshot of how this hierarchy looks on the LaminHub Sheets sidebar that also has wetlab things in a different "folder". This would then convey the integrated data management that closes the loop between wetlab & drylab.

Then, ideally within the callback, we do:

```python
ckpt_artifact.features.add_values("mlflow_run_id": mlflow_run_id)
```

**Comment by Zethson** [comment]
@falexwolf 

I had another stab at it and I think that we're converging on something good. I can optimize the screenshots further to make it clearer how MLflow runs and LaminDB runs are connected but before I spend more time on this I think it might be good to have another round of feedback of yours.

I understand the importance of this guide currently but I think that we should stop at a reasonable point before optimizing it further as we might overhaul it completely in a few weeks.

> image
> 
> I think this has no context up there because people don't yet understand what "tracking" even means, also because of the confusion between mlflow tracking and lamindb tracking. I'd move this box to the bottom of the guide.

I am not sure about that. This is the MLflow guide and MLflow has [tracking as a concept](https://mlflow.org/docs/latest/ml/tracking/). I expect people that read this guide to not be confused here actually. I further think that people that are reading this guide must have read the LaminDB quickstart anyways or they won't understand what's happening. Or do you think that it should be completely self-contained?
We added this box here initially to make it more obvious why we're defining a few parameters at the top which is rather uncommon.

**Comment by Zethson** [comment]
> Can we get the mlflow run id from self: Callback?

No. It's not the scope of the trainer.

> Then, ideally within the callback, we do:
> `ckpt_artifact.features.add_values("mlflow_run_id": mlflow_run_id)`

We're already doing that but the user has to pass the `mlflow_run_id`.

> For managing MLFlow runs, I recommend using Record, because we'll now have the ability to take notes etc. and it should look very intuitive on the GUI very soon (more intuitive than ulabel).
> All of this only has to be defined once and in the future will likely often be done on the UI:
> `run_type = ln.Record(name="MLFlow runs", is_type=True).save()`
> `my_runs_this_week = ln.Record(name="My runs week 3", is_type=True, type=mlflow_type).save()`
> `ln.Feature(name="mlflow_run_id", dtype=run_type).save()`

Blows it up but

> What gets really scary now is the "define_features_and_labels()` part.
> 
> To this end we could add `ln.integrations.create_mlflow_schema()` similar to what we have here: https://docs.lamin.ai/lamindb.examples.schemas
> 
> And similar to `import_source()`. Just something that makes all these MLFlow-related types and features with one line.

I'll see what I can do here. It should be in `ln.examples` like CxG.

**Comment by falexwolf** [comment]
Discuss? https://us06web.zoom.us/j/6174652065?pwd=cTdtKzhRd1p0ZjRoU2FscHNpU3lydz09
